src/libraries/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -91,5 +123,99 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
 
             return default;
         }
+
+        /// <summary>Returns the element at a specified index in a sequence or a default value if the index is out of range.</summary>
+        /// <param name="source">An <see cref="IEnumerable{T}" /> to return an element from.</param>
+        /// <param name="index">The index of the element to retrieve, which is either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        /// <returns>
+        ///   <see langword="default" /> if <paramref name="index" /> is outside the bounds of the <paramref name="source" /> sequence; otherwise, the element at the specified position in the <paramref name="source" /> sequence.
+        /// </returns>
+        public static TSource? ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (!index.IsFromEnd)
+            {
+                return source.ElementAtOrDefault(index.Value);
+            }
+
+            TryGetElementAt(source, index.Value, out TSource? result);
+            return result;
+        }
+
+        private static bool TryGetElementAt<TSource>(IEnumerable<TSource> source, int indexFromEnd, [NotNullWhen(true)] out TSource? result)
+        {
+            Debug.Assert(indexFromEnd >= 0);
+
+            result = default;
+            if (indexFromEnd == 0)
+            {
+                return false;
+            }
+
+            if (source is IPartition<TSource> partition)
+            {
+                int count = partition.GetCount(onlyIfCheap: true);
+                if (count == 0)
+                {
+                    return false;
+                }
+
+                if (count > 0)
+                {
+                    if (indexFromEnd <= count)
+                    {
+                        result = partition.TryGetElementAt(count - indexFromEnd, out bool found);
+                        return found;
+                    }
+
+                    return false;
+                }
+            }
+            else if (source is IList<TSource> list)
+            {
+                int count = list.Count;
+                if (indexFromEnd <= count)
+                {
+                    result = list[count - indexFromEnd]!;
```

Comment:
This ! shouldn't be necessary once the attribute is fixed per my previous comment.

Reply:
Updated.